docs/markdown/Python/python/python-lockfiles.md

```diff
@@ -0,0 +1,271 @@
+
+When used naively, this dependency resolution process is unstable: if you run a resolve, and then some time later run another resolve on the same inputs, you may end up with a different resulting set of dependencies. This is because new versions of direct or transitive dependencies may have been published (or, in rare cases, yanked) between the two runs.
+
+This is an issue both for correctness (your code may not be compatible with the new versions) and security (a new version may contain a vulnerability). A further security concern is that repeatedly downloading even the _same_ versions exposes you to greater risk if one of those versions is later compromised.
+
+Dependency resolution can also be a performance bottleneck, with the same complex version compatibility logic running repeatedly, and unnecessarily.
+
+Pants offers a solution to these issues that ensures stable, hermetic, secure builds over time, in the form of _lockfiles_.
+
+### What are lockfiles?
+
+A lockfile is a metadata file that enumerates specific pinned versions of every transitive third-party dependency. It also provides the expected SHA256 hashes of the downloadable artifacts (sdists and wheels) for each dependency. A lockfile can contain dependency version information that is valid across multiple platforms and Python interpreter versions. Lockfiles can be large and complex, but fortunately Pants will generate them for you!
+
+If you use lockfiles, and we highly recommend that you do, then Pants will use the locked transitive dependency versions in every build, and only change them when you deliberately update your lockfiles. Pants will also verify the downloaded artifacts against their expected hashes, to ensure that they haven't been compromised after the lockfile was generated.
+
+Pants supports multiple lockfiles for different parts of your repo, via the mechanism of "resolves" - logical names given to lockfiles so that they are easy to reference.
+
+> 📘 Lockfiles are generated by Pex
+>
+> Pants delegates lockfile creation and consumption to the [Pex](https://github.com/pantsbuild/pex) tool. So you may see standard lockfiles referred to as "Pex-style" lockfiles.
+>
+
+### Getting started with resolves
+
+First, you'll need to turn on the resolves functionality for the repo:
+
+```toml pants.toml
+[python]
+enable_resolves = true
+```
+
+
+
+Initially, Pants will assume a single resolve named `python-default` which references a lockfile at `3rdparty/python/default.lock`. You can change the name of the default resolve, and/or the location of its lockfile, via:
```

Comment:
It seems a bit off to have the new requirement being ICs must be specified and yet leading the narrative down a path of initially generating the lock without them. Poetry also requires ICs: https://python-poetry.org/docs/pyproject/#dependencies-and-dependency-groups; so it does not seem cruel or unusual to force this concept in from the get go, if not explaining it fully, at least writing it in the initial example with maybe ~"more on this below".

Reply:
I'll mention that this uses the default ICs.